### PR TITLE
CVOC : Indexed field accuracy (Ontoportal integration)

### PR DESCRIPTION
**Pull request by luddaniel**
Following agreements on next steps of #10145

**What this PR does / why we need it**:

This PR allows admins to specify which indexed fields must be use for a [:CVocConf](https://guides.dataverse.org/en/6.3/installation/config.html#cvocconf) configured service.

**Suggestions on how to test this**:

Related documentation : https://github.com/gdcc/dataverse-external-vocab-support/pull/21 
Ontoportal CVocConf example : [cvoc-agroportal-ontoportal-conf.json](https://github.com/Recherche-Data-Gouv/dataverse-external-vocab-support/blob/poc_ontoportal/examples/config/cvoc-agroportal-ontoportal-conf.json)

**Default behaviour (with `indexIn`)**:

Screenshot from 2024-04-18 16-31-45

**With `indexIn`**:

Screenshot from 2024-04-18 16-28-38

**Does this PR introduce a user interface change?**
Indirectly, it changes search result snippet.

### CODE REVIEW

File: doc/release-notes/9276-doc-cvoc-index-in.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+## Release Highlights
+
+### Updates on Support for External Vocabulary Services
+
+#### Indexed field accuracy
```
Comment (qqmyers) [inline doc/release-notes/9276-doc-cvoc-index-in.md:5]: Suggested change: 
### Improved Indexing with Compound Fields

When using an external vocabulary service with compound fields, you can now specify which field(s) will include additional indexed information, such as translations of an entry into other languages. This is done by adding the  `indexIn` in `retrieval-filtering`.
For more information, please check [GDCC/dataverse-external-vocab-support documentation](https://github.com/gdcc/dataverse-external-vocab-support/tree/main/docs).

File: src/main/java/edu/harvard/iq/dataverse/DatasetFieldServiceBean.java [hunk 2]
```diff
@@ -321,14 +322,15 @@ public Map<Long, JsonObject> getCVocConf(boolean byTermUriField){
                               + jo.getString("term-uri-field"));
                         }
                     }
-                    if (jo.containsKey("child-fields")) {
-                        JsonArray childFields = jo.getJsonArray("child-fields");
-                        for (JsonString elm : childFields.getValuesAs(JsonString.class)) {
-                            dft = findByNameOpt(elm.getString());
-                            logger.info("Found: " + dft.getName());
+                    if (jo.containsKey("managed-fields")) {
```
Comment (qqmyers) [inline src/main/java/edu/harvard/iq/dataverse/DatasetFieldServiceBean.java:325]: Just confirming - this was a bug? - the schema requires managed-fields these days

File: src/main/java/edu/harvard/iq/dataverse/DatasetFieldServiceBean.java [hunk 3]
```diff
@@ -321,14 +322,15 @@ public Map<Long, JsonObject> getCVocConf(boolean byTermUriField){
...
+                        JsonObject managedFields = jo.getJsonObject("managed-fields");
+                        for (String s : managedFields.keySet()) {
+                            dft = findByNameOpt(managedFields.getString(s));
                             if (dft == null) {
                                 logger.warning("Ignoring External Vocabulary setting for non-existent child field: "
-                                  + elm.getString());
+                                        + managedFields.getString(s));
+                            } else {
+                                logger.info("Found: " + dft.getName());
```
Comment (qqmyers) [inline src/main/java/edu/harvard/iq/dataverse/DatasetFieldServiceBean.java:333]: change logger.info() to fine()

File: src/main/java/edu/harvard/iq/dataverse/DatasetFieldServiceBean.java [hunk 4]
```diff
@@ -366,38 +368,48 @@ public void registerExternalVocabValues(DatasetField df) {
...
-     * Retrieves indexable strings from a cached externalvocabularyvalue entry.
-     * 
-     * This method assumes externalvocabularyvalue entries have been filtered and
-     * the externalvocabularyvalue entry contain a single JsonObject whose "personName" or "termName" values
-     * are either Strings or an array of objects with "lang" and ("value" or "content") keys. The
-     * string, or the "value/content"s for each language are added to the set.
-     * 
+     * Retrieves indexable strings from a cached externalvocabularyvalue entry filtered through retrieval-filtering configuration.
+     * <p>
+     * This method externalvocabularyvalue entries have been filtered and contains a single JsonObject.
```
Comment (qqmyers) [inline src/main/java/edu/harvard/iq/dataverse/DatasetFieldServiceBean.java:375]: This method assumes externalvocabularyvalue entries have been filtered and that they contain a single JsonObject.

File: src/main/java/edu/harvard/iq/dataverse/DatasetFieldServiceBean.java [hunk 5]
```diff
@@ -366,38 +368,48 @@ public void registerExternalVocabValues(DatasetField df) {
...
+     * Is handled : Strings, Array of Objects with "lang" and ("value" or "content") keys, Object with Strings as value or Object with Array of Strings as value.
```
Comment (qqmyers) [inline src/main/java/edu/harvard/iq/dataverse/DatasetFieldServiceBean.java:376]: Suggested change: Is handled : - > Cases Handled :

File: src/main/java/edu/harvard/iq/dataverse/DatasetFieldServiceBean.java [hunk 6]
```diff
@@ -366,38 +368,48 @@ public void registerExternalVocabValues(DatasetField df) {
...
+     * The string, or the "value/content"s for each language are added to the set.
```
Comment (qqmyers) [inline src/main/java/edu/harvard/iq/dataverse/DatasetFieldServiceBean.java:377]: Suggested change:
The string(s), or ...

File: src/main/java/edu/harvard/iq/dataverse/DatasetFieldServiceBean.java [hunk 7]
```diff
@@ -366,38 +368,48 @@ public void registerExternalVocabValues(DatasetField df) {
...
+     * This method can retrieve string values to be indexed in term-uri-field (parameter defined in CVOC configuration) or in "indexIn" field (optional parameter of retrieval-filtering defined in CVOC configuration).
```
Comment (qqmyers) [inline src/main/java/edu/harvard/iq/dataverse/DatasetFieldServiceBean.java:378]: Suggested change: 
Retrieved string values are indexed in the term-uri-field (parameter defined in CVOC configuration) by default, or in the field specified by an optional "indexIn" parameter in the retrieval-filtering defined in the CVOC configuration).

**Review by qqmyers — CHANGES_REQUESTED** [review]
Looks good overall. I suggested some doc changes and noted one place where a logger.info should be dropped to a fine(). Otherwise the code and tests looks good. 

W.r.t. docs - across all of these PR, it seems like adding 'services based on Ontoportal' to the list of SkosMos and ORCID in https://guides.dataverse.org/en/latest/admin/metadatacustomization.html#using-external-vocabulary-services would be good. Maybe adding 'how indexing is handled' to the sentence starting with "Configuration involves " as well.

File: src/main/java/edu/harvard/iq/dataverse/DatasetFieldServiceBean.java (hunk 2 above)
Comment (luddaniel) [reply src/main/java/edu/harvard/iq/dataverse/DatasetFieldServiceBean.java:325]: Not a bug, I would say old code, `child-fields` does not exists anymore so I replaced it with `managed-fields`. I think this code is not useful but at least the key is not obsolete.

**Review by luddaniel — COMMENTED** [review]
All documentation suggestions had been taken in account.
Updates in https://guides.dataverse.org/en/latest/admin/metadatacustomization.html#using-external-vocabulary-services are based on already updated documentation by @pdurbin : https://github.com/IQSS/dataverse/pull/10331/files#diff-618707003786b24e13bfefd16a45053ba610018723b3688d5edadf89fd49ce56

File: src/main/java/edu/harvard/iq/dataverse/DatasetFieldServiceBean.java (hunk 2 above)
Comment (qqmyers) [reply src/main/java/edu/harvard/iq/dataverse/DatasetFieldServiceBean.java:325]: Right - looks like it is just for logging purposes/to warn about incorrect field choices.

**Review by qqmyers — APPROVED** [review]
Looks good.

**Comment by sekmiller** [comment]
@luddaniel can you please update this branch with the latest from develop. There are merge conflicts. Thanks!

**Pull request merged**